File: lib/src/features/monitoring/presentation/widgets/quality_colors.dart
import 'package:flutter/material.dart';

import '../../domain/alert.dart';
import '../../domain/machine_state.dart';
import '../../domain/quality_evaluation.dart';
import '../../domain/quality_spec.dart';

/// Green that reads as "good" in both themes without borrowing the seed colour,
/// which is amber and would be mistaken for a warning.
const Color _acceptable = Color(0xFF2E7D32);
const Color _warning = Color(0xFFE07B00);

/// Maps a grading outcome onto theme colours.
///
/// Status is never signalled by colour alone — every place these are used also
/// carries an icon and a text label, so the reading survives colour blindness
/// and a washed-out screen in the field.
extension QualityStatusPresentation on QualityStatus {
  Color color(ColorScheme scheme) => switch (this) {
        QualityStatus.acceptable => _acceptable,
        QualityStatus.warning => _warning,
        QualityStatus.outOfRange => scheme.error,
        QualityStatus.unrated => scheme.onSurfaceVariant,
        QualityStatus.missing => scheme.outline,
      };

  Color containerColor(ColorScheme scheme) => switch (this) {
        QualityStatus.acceptable => _acceptable.withValues(alpha: 0.12),
        QualityStatus.warning => _warning.withValues(alpha: 0.14),
        QualityStatus.outOfRange =>
          scheme.errorContainer.withValues(alpha: 0.55),
        QualityStatus.unrated => scheme.surfaceContainerHighest,
        QualityStatus.missing => scheme.surfaceContainerHighest,
      };

  IconData get icon => switch (this) {
        QualityStatus.acceptable => Icons.check_circle_outline,
        QualityStatus.warning => Icons.warning_amber_outlined,
        QualityStatus.outOfRange => Icons.error_outline,
        QualityStatus.unrated => Icons.info_outline,
        QualityStatus.missing => Icons.hourglass_empty,
      };

  /// Short form for chips and cards.
  String get chipLabel => switch (this) {
        QualityStatus.acceptable => 'In range',
        QualityStatus.warning => 'Warning',
        QualityStatus.outOfRange => 'Out of range',
        QualityStatus.unrated => 'Not graded',
        QualityStatus.missing => 'No data',
      };
}

/// Presentation for the batch-level verdict of §8.
extension QualityAssessmentPresentation on QualityAssessment {
  QualityStatus get asStatus => switch (this) {
        QualityAssessment.acceptable => QualityStatus.acceptable,
        QualityAssessment.requiresAttention => QualityStatus.warning,
        QualityAssessment.outsideParameters => QualityStatus.outOfRange,
        QualityAssessment.incomplete => QualityStatus.missing,
      };

  Color color(ColorScheme scheme) => asStatus.color(scheme);

  Color containerColor(ColorScheme scheme) => asStatus.containerColor(scheme);

  IconData get icon => switch (this) {
        QualityAssessment.acceptable => Icons.verified_outlined,
        QualityAssessment.requiresAttention => Icons.warning_amber_outlined,
        QualityAssessment.outsideParameters => Icons.report_problem_outlined,
        QualityAssessment.incomplete => Icons.hourglass_empty,
      };
}

extension BatchRecommendationPresentation on BatchRecommendation {
  IconData get icon => switch (this) {
        BatchRecommendation.readyForStorage => Icons.inventory_2_outlined,
        BatchRecommendation.additionalFiltration => Icons.filter_alt_outlined,
        BatchRecommendation.holdForReview => Icons.pan_tool_outlined,
        BatchRecommendation.awaitingData => Icons.more_horiz,
      };
}

extension AlertSeverityPresentation on AlertSeverity {
  Color color(ColorScheme scheme) => switch (this) {
        AlertSeverity.info => scheme.primary,
        AlertSeverity.warning => _warning,
        AlertSeverity.critical => scheme.error,
      };
}

extension AlertKindPresentation on AlertKind {
  IconData get icon => switch (this) {
        AlertKind.parameterOutOfRange => Icons.error_outline,
        AlertKind.parameterWarning => Icons.warning_amber_outlined,
        AlertKind.sensorFault => Icons.sensors_off_outlined,
        AlertKind.connectionLost => Icons.bluetooth_disabled_outlined,
        AlertKind.machineError => Icons.report_problem_outlined,
        AlertKind.batchCompleted => Icons.task_alt_outlined,
        AlertKind.batchStarted => Icons.play_circle_outline,
      };
}

extension MachineStatusPresentation on MachineStatus {
  Color color(ColorScheme scheme) => switch (this) {
        MachineStatus.running => _acceptable,
        MachineStatus.completed => _acceptable,
        MachineStatus.paused => _warning,
        MachineStatus.error => scheme.error,
        MachineStatus.ready => scheme.primary,
        MachineStatus.connected => scheme.onSurfaceVariant,
        MachineStatus.disconnected => scheme.outline,
      };

  IconData get icon => switch (this) {
        MachineStatus.running => Icons.play_circle_outline,
        MachineStatus.completed => Icons.check_circle_outline,
        MachineStatus.paused => Icons.pause_circle_outline,
        MachineStatus.error => Icons.error_outline,
        MachineStatus.ready => Icons.power_settings_new,
        MachineStatus.connected => Icons.link,
        MachineStatus.disconnected => Icons.link_off,
      };
}

extension SensorParameterPresentation on SensorParameter {
  IconData get icon => switch (this) {
        SensorParameter.ph => Icons.science_outlined,
        SensorParameter.moisture => Icons.water_drop_outlined,
        SensorParameter.temperature => Icons.thermostat_outlined,
        SensorParameter.electricalConductivity => Icons.bolt_outlined,
        SensorParameter.turbidity => Icons.blur_on_outlined,
        SensorParameter.color => Icons.palette_outlined,
        SensorParameter.weight => Icons.scale_outlined,
        SensorParameter.flow => Icons.waves_outlined,
      };
}

File: react_native/src/features/monitoring/presentation/widgets/qualityColors.ts
import type { IconName } from '@/core/components/Icon';
import { alpha, statusColors, type ColorScheme } from '@/core/theme/theme';

import type { AlertKind, AlertSeverity } from '../../domain/alert';
import type { MachineStatus } from '../../domain/machineState';
import type { BatchRecommendation, QualityAssessment } from '../../domain/qualityEvaluation';
import type { QualityStatus, SensorParameter } from '../../domain/qualitySpec';

/**
 * Maps a grading outcome onto theme colours.
 *
 * Status is never signalled by colour alone — every place these are used also
 * carries an icon and a text label, so the reading survives colour blindness
 * and a washed-out screen in the field.
 */
export function statusColor(status: QualityStatus, scheme: ColorScheme): string {
  switch (status) {
    case 'acceptable':
      return statusColors.acceptable;
    case 'warning':
      return statusColors.warning;
    case 'outOfRange':
      return scheme.error;
    case 'unrated':
      return scheme.onSurfaceVariant;
    case 'missing':
      return scheme.outline;
  }
}

export function statusContainerColor(status: QualityStatus, scheme: ColorScheme): string {
  switch (status) {
    case 'acceptable':
      return alpha(statusColors.acceptable, 0.12);
    case 'warning':
      return alpha(statusColors.warning, 0.14);
    case 'outOfRange':
      return alpha(scheme.errorContainer, 0.55);
    case 'unrated':
    case 'missing':
      return scheme.surfaceContainerHighest;
  }
}

export function statusIcon(status: QualityStatus): IconName {
  switch (status) {
    case 'acceptable':
      return 'check-circle-outline';
    case 'warning':
      return 'warning-amber';
    case 'outOfRange':
      return 'error-outline';
    case 'unrated':
      return 'info-outline';
    case 'missing':
      return 'hourglass-empty';
  }
}

/** Short form for chips and cards. */
export function statusChipLabel(status: QualityStatus): string {
  switch (status) {
    case 'acceptable':
      return 'In range';
    case 'warning':
      return 'Warning';
    case 'outOfRange':
      return 'Out of range';
    case 'unrated':
      return 'Not graded';
    case 'missing':
      return 'No data';
  }
}

/** Presentation for the batch-level verdict of §8. */
export function assessmentAsStatus(assessment: QualityAssessment): QualityStatus {
  switch (assessment) {
    case 'acceptable':
      return 'acceptable';
    case 'requiresAttention':
      return 'warning';
    case 'outsideParameters':
      return 'outOfRange';
    case 'incomplete':
      return 'missing';
  }
}

export const assessmentColor = (a: QualityAssessment, scheme: ColorScheme) =>
  statusColor(assessmentAsStatus(a), scheme);

export const assessmentContainerColor = (a: QualityAssessment, scheme: ColorScheme) =>
  statusContainerColor(assessmentAsStatus(a), scheme);

export function assessmentIcon(assessment: QualityAssessment): IconName {
  switch (assessment) {
    case 'acceptable':
      return 'verified';
    case 'requiresAttention':
      return 'warning-amber';
    case 'outsideParameters':
      return 'report-problem';
    case 'incomplete':
      return 'hourglass-empty';
  }
}

export function recommendationIcon(recommendation: BatchRecommendation): IconName {
  switch (recommendation) {
    case 'readyForStorage':
      return 'inventory-2';
    case 'additionalFiltration':
      return 'filter-alt';
    case 'holdForReview':
      return 'pan-tool';
    case 'awaitingData':
      return 'more-horiz';
  }
}

export function severityColor(severity: AlertSeverity, scheme: ColorScheme): string {
  switch (severity) {
    case 'info':
      return scheme.primary;
    case 'warning':
      return statusColors.warning;
    case 'critical':
      return scheme.error;
  }
}

export function alertKindIcon(kind: AlertKind): IconName {
  switch (kind) {
    case 'parameterOutOfRange':
      return 'error-outline';
    case 'parameterWarning':
      return 'warning-amber';
    case 'sensorFault':
      return 'sensors-off';
    case 'connectionLost':
      return 'bluetooth-disabled';
    case 'machineError':
      return 'report-problem';
    case 'batchCompleted':
      return 'task-alt';
    case 'batchStarted':
      return 'play-circle-outline';
  }
}

export function machineStatusColor(status: MachineStatus, scheme: ColorScheme): string {
  switch (status) {
    case 'running':
    case 'completed':
      return statusColors.acceptable;
    case 'paused':
      return statusColors.warning;
    case 'error':
      return scheme.error;
    case 'ready':
      return scheme.primary;
    case 'connected':
      return scheme.onSurfaceVariant;
    case 'disconnected':
      return scheme.outline;
  }
}

export function machineStatusIcon(status: MachineStatus): IconName {
  switch (status) {
    case 'running':
      return 'play-circle-outline';
    case 'completed':
      return 'check-circle-outline';
    case 'paused':
      return 'pause-circle-outline';
    case 'error':
      return 'error-outline';
    case 'ready':
      return 'power-settings-new';
    case 'connected':
      return 'link';
    case 'disconnected':
      return 'link-off';
  }
}

export function parameterIcon(parameter: SensorParameter): IconName {
  switch (parameter) {
    case 'ph':
      return 'science';
    case 'moisture':
      return 'water-drop';
    case 'temperature':
      return 'thermostat';
    case 'electricalConductivity':
      return 'bolt';
    case 'turbidity':
      return 'blur-on';
    case 'color':
      return 'palette';
    case 'weight':
      return 'scale';
    case 'flow':
      return 'waves';
  }
}
